Tahi-project/tahi
# APERTA-10431 Prevent default permissions on Custom Tasks

JIRA issue: https://jira.plos.org/jira/browse/APERTA-10431

#### What this PR does:

The new card permission system works by adding a `filter_by_card_id` column to the `permissions` table of the database. When the system is determining whether a given user can `view` (etc.) a task, this `filter_by_card_id` column will be taken into account. If the value is `null`, it applies to all tasks. If it is not-`null`, it only applies to those tasks that have a `card_version_id` that has a `card_id` that is equal to the value of the `filter_by_card_id` column. This restricts that users access to a particular custom card.

Due to the way our roles and permissions were being seeded, some permissions were generated that had an `applies_to` column with the value `CustomCardType` and a `NULL` `filter_by_card_id`. This meant that those users could perform those actions on all custom cards. This change fixes that.

#### Special instructions for Review or PO:

To test:
1. Login as site admin
2. Create a custom card with an input field [1] and publish it
3. Set no permissions
4. Add the card to a paper
5. Login as a staff admin
6. Expected result: staff admin can neither view nor edit the task
7. As site admin: Add a view permission to the custom card
8. Result: staff admin can view (but not change) the task
9. As site admin: Add the edit permission to the custom card
10. Result: staff admin can edit the task

You can try the same with the "handling editor" role

[1]
```xml
<?xml version="1.0" encoding="UTF-8"?>
<card required-for-submission="false" workflow-display-only="false">
  <content content-type="display-children">
    <content content-type="radio" value-type="boolean">
      <text>test?</text>
    </content>
  </content>
</card>
```

---

#### Code Review Tasks:

**Author tasks** (delete tasks that don't apply to your PR, this list should be finished before code review):


If I need to migrate existing data:
- [X] If there are steps to take outside of `rake db:migrate` for Heroku or other environments, I added copy-pastable instructions to [the confluence release page](https://developer.plos.org/confluence/display/TAHI/Deployment+information+for+Release)
- [ ] I verified the data-migration's results on a copy of production data (complicated migrations should also have real specs)
- [ ] I've talked through the ramifications of the data-migration with Product Owners in regards to deployment timing
- [X] If I created a data migration, I added pre- and post-migration assertions.

**Reviewer tasks** (these should be checked or somehow noted before passing on to PO):
